src/scheduler.cc

```diff
@@ -442,27 +451,34 @@ void SchedulerService::perform_pulls() {
     const std::pair<WorkerId, ObjRef>& pull = pull_queue_[i];
     ObjRef objref = pull.second;
     WorkerId workerid = pull.first;
+    ObjStoreId objstoreid = get_store(workerid);
     if (!has_canonical_objref(objref)) {
       RAY_LOG(RAY_ALIAS, "objref " << objref << " does not have a canonical_objref, so continuing");
       continue;
     }
     ObjRef canonical_objref = get_canonical_objref(objref);
     RAY_LOG(RAY_DEBUG, "attempting to pull objref " << pull.second << " with canonical objref " << canonical_objref << " to objstore " << get_store(workerid));
-
-    objtable_lock_.lock();
-    int num_stores = objtable_[canonical_objref].size();
-    objtable_lock_.unlock();
-
+    int num_stores;
+    {
+      std::lock_guard<std::mutex> objects_lock(objects_lock_);
+      num_stores = objtable_[canonical_objref].size();
+    }
     if (num_stores > 0) {
+      bool object_present_or_in_transit;
       {
-        std::lock_guard<std::mutex> objtable_lock(objtable_lock_);
-        if (!std::binary_search(objtable_[canonical_objref].begin(), objtable_[canonical_objref].end(), get_store(workerid))) {
-          // The worker's local object store does not already contain objref, so ship
-          // it there from an object store that does have it.
-          ObjStoreId objstoreid = pick_objstore(canonical_objref);
-          deliver_object(canonical_objref, objstoreid, get_store(workerid));
+        std::lock_guard<std::mutex> objects_lock(objects_lock_);
+        bool object_present = std::binary_search(objtable_[canonical_objref].begin(), objtable_[canonical_objref].end(), objstoreid);
```

Comment:
You've referenced `objtable_[canonical_objref]` a dozen times here, just assign it to a variable and use that variable instead.